cognite/neat/rules/analysis/_base.py

```diff
@@ -1,13 +1,392 @@
+T_PropertyEntity = TypeVar("T_PropertyEntity", bound=Entity | str)
+
+
+@dataclass(frozen=True)
+class Linkage(Generic[T_ClassEntity, T_PropertyEntity]):
+    source_class: T_ClassEntity
+    connecting_property: T_PropertyEntity
+    target_class: T_ClassEntity
+    max_occurrence: int | float | None
+
+
+class LinkageSet(set, Generic[T_ClassEntity, T_PropertyEntity], Set[Linkage[T_ClassEntity, T_PropertyEntity]]):
+    @property
+    def source_class(self) -> set[T_ClassEntity]:
+        return {link.source_class for link in self}
+
+    @property
+    def target_class(self) -> set[T_ClassEntity]:
+        return {link.target_class for link in self}
+
+    def get_target_classes_by_source(self) -> dict[T_ClassEntity, set[T_ClassEntity]]:
+        target_classes_by_source: dict[T_ClassEntity, set[T_ClassEntity]] = defaultdict(set)
+        for link in self:
+            target_classes_by_source[link.source_class].add(link.target_class)
+        return target_classes_by_source
+
+    def to_pandas(self) -> pd.DataFrame:
+        # Todo: Remove this method
+        return pd.DataFrame(
+            [
+                {
+                    "source_class": link.source_class,
+                    "connecting_property": link.connecting_property,
+                    "target_class": link.target_class,
+                    "max_occurrence": link.max_occurrence,
+                }
+                for link in self
+            ]
+        )
+
+
+class BaseAnalysis(ABC, Generic[T_Rules, T_Class, T_Property, T_ClassEntity, T_PropertyEntity]):
+    def __init__(self, rules: T_Rules) -> None:
+        self.rules = rules
+
+    @abstractmethod
+    def _get_classes(self) -> list[T_Class]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_properties(self) -> list[T_Property]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_reference(self, class_or_property: T_Class | T_Property) -> ReferenceEntity | None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_cls_entity(self, class_: T_Class | T_Property) -> T_ClassEntity:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_prop_entity(self, property_: T_Property) -> T_PropertyEntity:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_cls_parents(self, class_: T_Class) -> list[T_ClassEntity] | None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_reference_rules(self) -> T_Rules | None:
+        raise NotImplementedError
+
+    @classmethod
+    @abstractmethod
+    def _set_cls_entity(cls, property_: T_Property, class_: T_ClassEntity) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_object(self, property_: T_Property) -> T_ClassEntity | None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_max_occurrence(self, property_: T_Property) -> int | float | None:
+        raise NotImplementedError
```

Comment:
These can be moved into the rules, property and class classes. But kept it here to avoid doing too many changes in this round.